File: src-tauri/src/state.rs
use crate::database::Database;
use crate::models::{AppConfig, ProcessInfo};
use crate::platform::{create_platform_manager, PlatformProcessManager};
use portable_pty::MasterPty;
use std::collections::HashMap;
use std::io::{BufRead, Write};
use std::path::PathBuf;
use std::sync::Mutex;
use tokio::process::Child;

pub struct ManagedProcess {
    pub child: Child,
    pub manually_stopped: bool,
    pub session_id: Option<String>,
    /// Group ID this process belongs to
    pub group_id: String,
    /// PTY master for interactive processes (optional)
    pub pty_master: Option<Box<dyn MasterPty + Send>>,
    /// PTY writer for sending input to interactive processes (optional)
    pub pty_writer: Option<Box<dyn Write + Send>>,
    /// Whether this process uses PTY for interactive mode
    pub is_interactive: bool,
    /// The actual process ID (for PTY processes, this is different from child.id())
    pub real_pid: Option<u32>,
}

/// Size of the PTY terminal (rows x columns)
#[derive(Debug, Clone, Copy, Default)]
pub struct PtyDimensions {
    pub rows: u16,
    pub cols: u16,
}

impl PtyDimensions {
    pub fn new(rows: u16, cols: u16) -> Self {
        Self { rows, cols }
    }
}

pub struct AppState {
    /// In-memory configuration
    pub config: Mutex<AppConfig>,
    /// Unified database for all data
    pub database: Mutex<Database>,
    /// Path to the database file
    pub database_path: PathBuf,
    pub processes: Mutex<HashMap<String, ManagedProcess>>,
    pub process_infos: Mutex<HashMap<String, ProcessInfo>>,
    pub log_dir: PathBuf,
    /// Maps project_id to active session_id
    pub active_sessions: Mutex<HashMap<String, String>>,
    /// Path to the PID file for tracking running processes
    pub pid_file_path: PathBuf,
    /// Platform-specific process manager
    platform_manager: Box<dyn PlatformProcessManager>,
    /// YAML file watcher
    pub yaml_watcher: Mutex<crate::file_watcher::YamlWatcher>,
    /// Configuration directory path
    pub config_dir: PathBuf,
}

impl AppState {
    pub fn new(config: AppConfig, log_dir: PathBuf, config_dir: PathBuf) -> Self {
        let _ = std::fs::create_dir_all(&log_dir);
        let pid_file_path = config_dir.join("running_pids.txt");
        let platform_manager = Box::new(create_platform_manager());

        // Initialize unified database
        let database_path = config_dir.join("runner-ui.db");
        let database = Database::open(&database_path).expect("Failed to open database");

        Self {
            config: Mutex::new(config),
            database: Mutex::new(database),
            database_path: database_path.clone(),
            processes: Mutex::new(HashMap::new()),
            process_infos: Mutex::new(HashMap::new()),
            log_dir,
            active_sessions: Mutex::new(HashMap::new()),
            pid_file_path,
            platform_manager,
            yaml_watcher: Mutex::new(crate::file_watcher::YamlWatcher::new()),
            config_dir,
        }
    }

    pub fn log_file_path(&self, project_id: &str) -> PathBuf {
        self.log_dir.join(format!("{}.log", project_id))
    }

    /// Save a process ID to the PID file
    pub fn save_pid(&self, pid: u32) {
        if let Ok(mut file) = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.pid_file_path)
        {
            let _ = writeln!(file, "{}", pid);
        }
    }

    /// Remove a process ID from the PID file
    pub fn remove_pid(&self, pid: u32) {
        if let Ok(content) = std::fs::read_to_string(&self.pid_file_path) {
            let remaining: Vec<String> = content
                .lines()
                .filter(|line| line.trim().parse::<u32>().ok() != Some(pid))
                .map(|s| s.to_string())
                .collect();
            let _ = std::fs::write(&self.pid_file_path, remaining.join("\n"));
        }
    }

    /// Read all stored PIDs from the PID file
    pub fn read_stored_pids(&self) -> Vec<u32> {
        if let Ok(file) = std::fs::File::open(&self.pid_file_path) {
            let reader = std::io::BufReader::new(file);
            reader
                .lines()
                .filter_map(|line| line.ok()?.trim().parse::<u32>().ok())
                .collect()
        } else {
            Vec::new()
        }
    }

    /// Clear the PID file
    pub fn clear_pid_file(&self) {
        let _ = std::fs::write(&self.pid_file_path, "");
    }

    /// Get platform manager
    #[allow(dead_code)]
    pub fn platform(&self) -> &dyn PlatformProcessManager {
        self.platform_manager.as_ref()
    }

    /// Get the database (for convenience)
    pub fn db(&self) -> std::sync::MutexGuard<'_, Database> {
        self.database.lock().unwrap()
    }

    /// Get the database path
    pub fn db_path(&self) -> &PathBuf {
        &self.database_path
    }
}

impl Drop for AppState {
    fn drop(&mut self) {
        // Last-resort cleanup: kill all processes
        if let Ok(processes) = self.processes.lock() {
            for managed in processes.values() {
                let pid_to_kill = if managed.is_interactive {
                    managed.real_pid
                } else {
                    managed.child.id()
                };

                if let Some(pid) = pid_to_kill {
                    self.platform_manager.force_kill(pid);
                }
            }
        }

        // Also kill any orphaned processes from the PID file
        for pid in self.read_stored_pids() {
            self.platform_manager.force_kill(pid);
        }
        self.clear_pid_file();
    }
}

File: src-electron/main/services/state.ts
/**
 * Application state management for the Electron main process.
 * This is the equivalent of src-tauri/src/state.rs
 */

import { app } from 'electron';
import { join } from 'path';
import { existsSync, mkdirSync, readFileSync, writeFileSync, appendFileSync } from 'fs';
import type {
  AppConfig,
  ProcessInfo,
  ManagedProcess,
  Group,
} from '../../shared/types';
import { Database } from './database';

/**
 * Application state singleton
 */
class AppState {
  private static instance: AppState | null = null;

  // Configuration
  config: AppConfig = { groups: [] };

  // Database
  database!: Database;
  databasePath: string = '';

  // Process management
  processes: Map<string, ManagedProcess> = new Map();
  processInfos: Map<string, ProcessInfo> = new Map();
  activeSessions: Map<string, string> = new Map();

  // Paths
  logDir: string = '';
  configDir: string = '';
  pidFilePath: string = '';

  private constructor() {
    // Private constructor for singleton
  }

  /**
   * Get the singleton instance
   */
  static getInstance(): AppState {
    if (!AppState.instance) {
      AppState.instance = new AppState();
    }
    return AppState.instance;
  }

  /**
   * Initialize the application state
   */
  async initialize(): Promise<void> {
    // Set up paths
    this.configDir = join(app.getPath('userData'));
    this.logDir = join(this.configDir, 'logs');
    this.pidFilePath = join(this.configDir, 'running_pids.txt');
    this.databasePath = join(this.configDir, 'runner-ui.db');

    // Ensure directories exist
    if (!existsSync(this.configDir)) {
      mkdirSync(this.configDir, { recursive: true });
    }
    if (!existsSync(this.logDir)) {
      mkdirSync(this.logDir, { recursive: true });
    }

    // Initialize database
    this.database = new Database(this.databasePath);

    // Load configuration from database
    await this.loadConfig();

    // Kill any orphaned processes from previous runs
    this.killOrphanedProcesses();
  }

  /**
   * Load configuration from database
   */
  private async loadConfig(): Promise<void> {
    try {
      const groups = this.database.getAllGroups();
      this.config = { groups };
    } catch (error) {
      console.error('Failed to load config from database:', error);
      this.config = { groups: [] };
    }
  }

  /**
   * Reload configuration from database
   */
  async reloadConfig(): Promise<void> {
    await this.loadConfig();
  }

  /**
   * Get log file path for a project
   */
  logFilePath(projectId: string): string {
    return join(this.logDir, `${projectId}.log`);
  }

  /**
   * Save a PID to the PID file (for orphan cleanup on restart)
   */
  savePid(pid: number): void {
    try {
      appendFileSync(this.pidFilePath, `${pid}\n`);
    } catch (error) {
      console.error('Failed to save PID:', error);
    }
  }

  /**
   * Remove a PID from the PID file
   */
  removePid(pid: number): void {
    try {
      if (!existsSync(this.pidFilePath)) return;

      const content = readFileSync(this.pidFilePath, 'utf-8');
      const remaining = content
        .split('\n')
        .filter((line) => {
          const parsed = parseInt(line.trim(), 10);
          return !isNaN(parsed) && parsed !== pid;
        })
        .join('\n');

      writeFileSync(this.pidFilePath, remaining);
    } catch (error) {
      console.error('Failed to remove PID:', error);
    }
  }

  /**
   * Read all stored PIDs from the PID file
   */
  readStoredPids(): number[] {
    try {
      if (!existsSync(this.pidFilePath)) return [];

      const content = readFileSync(this.pidFilePath, 'utf-8');
      return content
        .split('\n')
        .map((line) => parseInt(line.trim(), 10))
        .filter((pid) => !isNaN(pid));
    } catch (error) {
      console.error('Failed to read PIDs:', error);
      return [];
    }
  }

  /**
   * Clear the PID file
   */
  clearPidFile(): void {
    try {
      writeFileSync(this.pidFilePath, '');
    } catch (error) {
      console.error('Failed to clear PID file:', error);
    }
  }

  /**
   * Kill any orphaned processes from previous app runs
   */
  private killOrphanedProcesses(): void {
    const pids = this.readStoredPids();
    for (const pid of pids) {
      try {
        // Check if process is still running and kill it
        process.kill(pid, 0); // Check if process exists
        process.kill(pid, 'SIGKILL'); // Kill it
      } catch {
        // Process doesn't exist or can't be killed
      }
    }
    this.clearPidFile();
  }

  /**
   * Shutdown cleanup
   */
  shutdown(): void {
    // Kill all managed processes
    for (const [projectId, managed] of this.processes) {
      try {
        const pid = managed.realPid ?? managed.child?.pid;
        if (pid) {
          process.kill(pid, 'SIGKILL');
        }
        if (managed.ptyProcess) {
          managed.ptyProcess.kill();
        }
      } catch {
        // Process already dead
      }
    }
    this.processes.clear();
    this.clearPidFile();

    // Close database
    if (this.database) {
      this.database.close();
    }
  }

  /**
   * Find a group by ID
   */
  findGroup(groupId: string): Group | undefined {
    return this.config.groups.find((g) => g.id === groupId);
  }

  /**
   * Find a project by ID across all groups
   */
  findProject(projectId: string): { group: Group; project: Group['projects'][0] } | undefined {
    for (const group of this.config.groups) {
      const project = group.projects.find((p) => p.id === projectId);
      if (project) {
        return { group, project };
      }
    }
    return undefined;
  }
}

// Singleton accessor functions
let stateInstance: AppState | null = null;

/**
 * Initialize the application state (call once at startup)
 */
export async function initializeState(): Promise<void> {
  stateInstance = AppState.getInstance();
  await stateInstance.initialize();
}

/**
 * Get the application state
 */
export function getState(): AppState {
  if (!stateInstance) {
    throw new Error('State not initialized. Call initializeState() first.');
  }
  return stateInstance;
}

/**
 * Shutdown the application state
 */
export function shutdownState(): void {
  if (stateInstance) {
    stateInstance.shutdown();
    stateInstance = null;
  }
}
